### PR TITLE
docs: PBS Scheduler Support [FOUNDENG-184]

### PR DESCRIPTION
**Pull request by jerryharrow**
## Description

This is the documentation for our PBS scheduler support as an
alternative to Slurm which is now fully functional.

## Checklist
- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.
- [ ] If modifying `/webui/react/src/shared/` verify `make -C webui/react test-shared` passes.

### CODE REVIEW

File: docs/reference/reference-deploy/config/master-config-reference.rst [hunk 1]
```diff
@@ -285,37 +285,44 @@ The master supports the following configuration settings:
...
 
-      -  ``partition_overrides``: A map of Slurm partition names to partition-level overrides. For
+         For PBS, the ``ngpus`` option can be used to identify the number of GPUs available on a
+         node.
+
+         For Slurm, ``GresTypes=gpu`` is set in the Slurm configuration, and nodes with GPUs have
+         properly configured GRES indicating the presence of any GPUs. The default is true. When
+         false, Determined will request ``slots_per_trial`` nodes and utilize only GPU 0 on each
+         node. It is the user's responsibility to ensure that GPUs will be available on nodes
+         selected for the job using other configurations such as targeting a specific resource pool
```
Comment (rcorujo) [inline docs/reference/reference-deploy/config/master-config-reference.rst:314]: Need a comma after "configurations".

File: docs/reference/reference-deploy/config/master-config-reference.rst [hunk 2]
```diff
@@ -285,37 +285,44 @@ The master supports the following configuration settings:
...
+         with only GPU nodes or specifying a Slurm constraint in the experiment configurSation.
```
Comment (rcorujo) [inline docs/reference/reference-deploy/config/master-config-reference.rst:315]: "configureSation" should be "configuration".

File: docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst [hunk 3]
```diff
@@ -33,5 +33,128 @@ The launcher has the following additional requirements on the installation node:
...
+
+      Given those options, the simplest path to enabling Determined job preemption is by including D
+      in the ``preemption_order``. You may include ``R`` in the ``preemption_order``, but it is
+      disabled for Determined jobs. You may include ``C`` to the ``preemption_order`` if you
+      additionally configure a checkpoint script. Refer to the PBS documentation for details. If you
+      choose to implement a checkpoint script, you may initiate a Determined checkpoint by sending a
+      ``SIGTERM`` signal to the Determined job. When a Determined job receives a ``SIGTERM``, it
+      begins a checkpoint and graceful shutdown. To prevent unnecessary loss of work, it is
+      recommended that you then wait at least one Determined ``scheduling_unit`` for the job to
+      complete after sending the ``SIGTERM``. If after that period of time the job has not yet
```
[the remark this answers is not in the document]
Comment (jerryharrow) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst:159]: > If after that period of time the job has not terminated, then send a ``SIGKILL`` to forcibly release all resources.

I don't think , after time is right.   The if clause continues to terminated which is where the comma is.

[the remark this answers is not in the document]
Comment (jerryharrow) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst:159]: I dropped the yet.

[the remark this answers is not in the document]
Comment (rcorujo) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst:159]: > > If after that period of time the job has not terminated, then send a `SIGKILL` to forcibly release all resources.
> 
> I don't think , after time is right. The if clause continues to terminated which is where the comma is.

Personal preference.  If I was speaking that sentence, I would pause after "If" and "time", much like I would pause after "terminated.

File: docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/overview.rst [hunk 4]
```diff
@@ -1,31 +1,34 @@
...
-   Slurm deployment applies to the Enterprise Edition.
+   Slurm/PBS deployment applies to the Enterprise Edition.
 
-Determined Slurm integration delegates all job scheduling and prioritization to the Slurm workload
-manager. This integration enables existing Slurm workloads and Determined workloads to coexist and
-Determined workloads to access all of the advanced capabilities of the Slurm workload manager.
+This document describes how Determined can be configured to utilize HPC cluster scheduling systems
+via the Determined HPC launcher. In this type of configuration, Determined delegates all job
+scheduling and prioritization to the HPC workload manager (either Slurm or PBS). This integration
+enables existing HPC workloads and Determined workloads to coexist and Determined workloads to
```
[the remark this answers is not in the document]
Comment (jerryharrow) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/overview.rst:28]: ?  Not seeing it.  You mean:

| In this type of configuration, Determined delegates all job scheduling and prioritization to the HPC workload manager (either Slurm or PBS).

should be: scheduling, and prioritization ?

[the remark this answers is not in the document]
Comment (jerryharrow) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/overview.rst:28]: Or 
| This integration enables existing HPC workloads and Determined workloads to coexist and Determined workloads to

should be: This integration enables existing HPC workloads and, Determined workloads to coexist and Determined workloads to

?

[the remark this answers is not in the document]
Comment (rcorujo) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/overview.rst:28]: I think I meant a comma "before" the second "and", sorry:

This integration enables existing HPC workloads and Determined workloads to coexist, and Determined workloads to access all of the advanced capabilities of the HPC workload manager.

[the remark this answers is not in the document]
Comment (jerryharrow) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/overview.rst:28]: between "coexist, and"?

[the remark this answers is not in the document]
Comment (jerryharrow) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/overview.rst:28]: Got it.  Though grammerly.com claims it is wrong to add.

[the remark this answers is not in the document]
Comment (rcorujo) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/overview.rst:28]: Ok, go with grammerly. I was old told that "and" should be used in the context of "this and that", which is not the case here.  In this case, it's being used as "also".

**Comment by phillip-gaisford** [comment]
experiment-config-reference could use an update too, or it could wait until we document slots_per_node.

File: docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst [hunk 5]
```diff
@@ -33,5 +33,128 @@ The launcher has the following additional requirements on the installation node:
...
+   specific resource pool with only GPU nodes, or specifying a Slurm constraint in the experiment
+   configuration.
+
+-  Ensure homogeneous Slurm partitions.
+
+   Determined maps Slurm partitions to Determined resource pools. It is recommended that the nodes
+   within a partition are homogeneous for Determined to effectively schedule GPU jobs.
+
+   -  A Slurm partition with GPUs is identified as a CUDA/ROCM resource pool. The type is inherited
+      from the ``resource_manager.slot_type`` configuration. It can be also be specified-per
```
Comment (rcorujo) [inline docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst:81]: I'm not sure if there needs to be a hyphen between "specified" and "per".

File: docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst [hunk 6]
```diff
@@ -33,5 +33,128 @@ The launcher has the following additional requirements on the installation node:
...
+   -  The Determined default resource pool is set to the Slurm default partition.
+
+-  Tune the Slurm configuration for Determined job preemption.
+
+   Slurm preempts jobs using signals. When a Determined job receives SIGTERM, it begins a checkpoint
+   and graceful shutdown. To prevent unnecessary loss of work, it is recommended to set ``GraceTime
+   (secs)`` high enough to permit the job to complete an entire Determined ``scheduling_unit``.
+
+   To enable GPU job preemption, use ``PreemptMode=REQUEUE`` or ``PreemptMode=REQUEUE``, because
+   ``PreemptMode=SUSPEND`` does not release GPUs so does not allow a higher-priority job to access
```
Comment (rcorujo) [inline docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst:95]: "does not release GPUs so does not allow" does not read well.

File: docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst [hunk 7]
```diff
@@ -33,5 +33,128 @@ The launcher has the following additional requirements on the installation node:
...
+   identify GPUs in this manner, specify the :ref:`pbs section <cluster-configuration-slurm>`
+   ``gres_supported`` option to ``false`` when configuring Determined, and it will then be the
+   user's responsibility to ensure that GPUs will be available on nodes selected for the job using
+   other configurations such as targeting a specific resource pool with only GPU nodes, or
+   specifying a PBS constraint in the experiment configuration.
+
+-  Ensure homogeneous PBS queues.
+
+   Determined maps PBS queues to Determined resource pools. It is recommended that the nodes within
+   a queue are homogeneous for Determined to effectively schedule GPU jobs.
```
Comment (rcorujo) [inline docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst:120]: "are homogeneous" or "be homogeneous"?

File: docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst [hunk 8]
```diff
@@ -33,5 +33,128 @@ The launcher has the following additional requirements on the installation node:
...
+
+   -  A PBS queue with no GPUs is identified as an AUX resource pool.
+
+   -  The Determined default resource pool is set to the PBS default queue.
+
+-  Tune the PBS configuration for Determined job preemption.
+
+   PBS supports a wide variety of criteria to trigger job preemption, and you may use any per your
+   system and job requirements. Once a job is identified for preemption, PBS supports four different
+   options for job preemption which are specified via the preemption_order scheduling parameter. The
```
Comment (rcorujo) [inline docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst:134]: Does "preemption_order" need the two back quotes like you use elsewhere in the document?

File: docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst [hunk 9]
```diff
@@ -33,5 +33,128 @@ The launcher has the following additional requirements on the installation node:
...
+         This requires a custom checkpoint script is added to PBS.
+
+      ``R`` - Requeue the job.
+         Determined does not support the re-queueing of a task. Determined jobs specify the ``-r n``
+         option to PBS to prevent this case.
+
+      ``D`` - Delete the job.
+         Determined jobs support this option without configuration.
+
+      Given those options, the simplest path to enabling Determined job preemption is by including D
```
Comment (rcorujo) [inline docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst:151]: "to enabling" or "to enable"?

Comment (rcorujo) [reply docs/cluster-setup-guide/deploy-cluster/sysadmin-deploy-on-slurm/slurm-requirements.rst:151]: Does the "D" at the end need the two back quotes?

File: docs/reference/reference-deploy/config/master-config-reference.rst [hunk 10]
```diff
@@ -285,37 +285,44 @@ The master supports the following configuration settings:
 
       -  ``job_storage_root``: The shared directory where job-related files will be stored. It is
          where the needed Determined executables are copied to when the experiment is run, as well
-         as where the Slurm ``sbatch`` script and log files are created. This directory must be
-         visible to the launcher and from the compute nodes.
+         as where the Slurm/PBS scripts and log files are created. This directory must be writable
+         by the launcher and accessible to the compute nodes to enable the logs to be written from
```
Comment (rcorujo) [inline docs/reference/reference-deploy/config/master-config-reference.rst:289]: Not sure of "enable the logs from the submitted batch jobs to be written" sounds better.

File: docs/reference/reference-deploy/config/master-config-reference.rst [hunk 11]
```diff
@@ -285,37 +285,44 @@ The master supports the following configuration settings:
...
-         resource pool with only GPU nodes or specifying a Slurm constraint in the experiment
-         configuration.
+      -  ``gres_supported``: Indicates if GPU resources are properly configured in the HPC workload
+         manager.
 
-      -  ``partition_overrides``: A map of Slurm partition names to partition-level overrides. For
+         For PBS, the ``ngpus`` option can be used to identify the number of GPUs available on a
+         node.
+
+         For Slurm, ``GresTypes=gpu`` is set in the Slurm configuration, and nodes with GPUs have
```
Comment (rcorujo) [inline docs/reference/reference-deploy/config/master-config-reference.rst:310]: The "and nodes with GPUs have properly configured GRES" part is confusing. The nodes themselves don't configure anything, nor is anything configured on the nodes.  The nodes do need the proper NVIDIA or ROCM binaries, libraries, and drivers, though.

Comment (jerryharrow) [reply docs/reference/reference-deploy/config/master-config-reference.rst:310]: Not node configuration directly, the gres for the node needs to be configured to reflect the GPUs on the nodes.

**Pull request merged**